File: Online-judge/leetcode/minimum-fuel-cost-to-report-to-the-capital.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/minimum-fuel-cost-to-report-to-the-capital/description/
 * Time complexity: O(n) 
 * Space complexity: O(n) 
 */

class Solution {
private:
    long long fuels;
    long long dfs(int node, int parent, vector<vector<int>>& citys, int& seats) {
        int representatives = 1;
        for (int city : citys[node]) {
            if (city != parent) {
                representatives += dfs(city, node, citys, seats);
            }
        }
        if (node != 0) {
            fuels += ceil((double) representatives / seats);
        }
        return representatives;
    }

public:
    long long minimumFuelCost(vector<vector<int>>& roads, int seats) {
        int n = roads.size() + 1;
        vector<vector<int>> graph(n);

        for (auto& road : roads) {
            graph[road[0]].push_back(road[1]);
            graph[road[1]].push_back(road[0]);
        }
        dfs(0, -1, graph, seats);
        return fuels;
    }
};

File: Online-judge/leetcode/minimum-fuel-cost-to-report-to-the-capital.py
from ast import List
import math
from collections import defaultdict

class Solution:
    def dfs(self, node: int, parent: int) -> int:
        representative: int = 1
        for childNode in self.citys[node]:
            if childNode != parent:
                representative += self.dfs(childNode, node)
        if node != 0:
            self.fuel += math.ceil(representative / self.seats)
        return representative

    def minimumFuelCost(self, roads: List[List[int]], seats: int) -> int:
        self.seats = seats
        self.citys = defaultdict(list)
        self.fuel = 0

        for road in roads:
            self.citys[road[0]].append(road[1])
            self.citys[road[1]].append(road[0])
        self.dfs(0, -1)
        return self.fuel
